Refactor Xthin data out of CNode

ptschip opened the pull request:
Take any xthin related data out of CNode.  Carry the xthin data, in memory, directly in the thinblock or xthinblock.  Furthermore store the xthinblock/thinblock's as smart pointers within a class CBlock_ThinRelay which itself is derived from CBlock().  This way way just can pass around a shared pointer to the CBlock_ThinRelay() which contains all the data  to reconstruct the block as well as the block itself.

Also store the block pointers in a new map so we can access them, in the cases where we need to re-request transactions and then finally reconstruct the block.

Once this PR gets approved I'll pull in the Graphene and Compact Block data as well but I don't want to take the time to do that until this PR is finalized.

ptschip committed:
Move nSizeThinBlock out of CNode

Encapsulate the data into CThinBlock and CXThinBlock

Also encapuslate the thinblock and xthinblock within CBlock. This way
we can eventually just pass around a smart pointer to CBlock and it will
have everything we need to reconstruct the block from the underlying
xthin and/or thinblock data.

ptschip committed:
Move thinBlockWaitingForTxns out of CNode

ptschip committed:
Move thinblock hashe vectors out of CNode

ptschip committed:
fix mapmissing in compactblocks and grapheneblocks

ptschip committed:
Move mapMissingTx out of CNode

ptschip committed:
Move nLocalThinBlockBytes out of CNode

ptschip committed:
Remove CBlock from CNode for thintype relay

Create a subclass from CBlock called CBlock_ThinType and use it

This subclass carries short term data in memory which is used in
the reconstruction of a block from a thinblock or xthinblock.

Rename ClearThinBlockData() to ClearThinBlockBytes()

sickpig commented on src/blockrelay/blockrelay_common.cpp:
```diff
@@ -245,3 +246,37 @@ void ThinTypeRelay::RequestBlock(CNode *pfrom, const uint256 &hash)
+        return iter->second.second;
+    }
+    else
+        return std::shared_ptr<CBlock_ThinRelay>{};
```
nit: curly branches around return statement.

sickpig commented on src/blockrelay/compactblock.cpp:
```diff
@@ -190,7 +190,7 @@ bool CompactBlock::process(CNode *pfrom, uint64_t nSizeCompactBlock)
     // to first create the full list of compactblock shortid hashes, in proper order.
     //
-    // Also, create the mapMissingTx from all the supplied tx's in the compact block
+    // Also, create the mapMissingCompactBlockTx from all the supplied tx's in the compact block
```
does this mean that current `dev` is using the same data structure to store missing transactions for all propagation types we support?

ptschip committed:
Clarify return statement and add a test

ptschip commented:
On 19/03/2019 6:58 a.m., Andrea Suisani wrote:
>
> *@sickpig* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/blockrelay/blockrelay_common.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1611#discussion_r266838805>:
>
> > +        std::make_pair(pfrom->GetId(), std::make_pair(pblock->GetBlockHeader().GetHash(), pblock)));
> +    return pblock;
> +}
> +
> +std::shared_ptr<CBlock_ThinRelay> ThinTypeRelay::GetBlockToReconstruct(CNode *pfrom)
> +{
> +    // Retrieve a current instance of a block being reconstructed. This is typically used
> +    // when we have received the response of a re-request for more transactions.
> +    LOCK(cs_reconstruct);
> +    auto iter = mapBlocksReconstruct.find(pfrom->GetId());
> +    if (iter != mapBlocksReconstruct.end())
> +    {
> +        return iter->second.second;
> +    }
> +    else
> +        return std::shared_ptr<CBlock_ThinRelay>{};
>
> nit: curly branches around return statement.
>
thanks, clarified the code and added a test.


> ------------------------------------------------------------------------
>
> In src/blockrelay/compactblock.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1611#discussion_r266895070>:
>
> > @@ -190,7 +190,7 @@ bool CompactBlock::process(CNode *pfrom, uint64_t nSizeCompactBlock)
>      // Because the list of shorttxids is not complete (missing the prefilled transaction hashes), we need
>      // to first create the full list of compactblock shortid hashes, in proper order.
>      //
> -    // Also, create the mapMissingTx from all the supplied tx's in the compact block
> +    // Also, create the mapMissingCompactBlockTx from all the supplied tx's in the compact block
>
> does this mean that current |dev| is using the same data structure to
> store missing transactions for all propagation types we support?
>
yes...fixed it here in the refactor.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1611#pullrequestreview-216094023>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0KTbV8QnGdEifB-0yzgn0zKS7sFUks5vYOz-gaJpZM4b6bF3>.
>

gandrewstone commented on src/primitives/block.h:
```diff
@@ -193,6 +195,26 @@ class CBlock : public CBlockHeader
 };
 
+// Used for thin type blocks that we want to reconstruct
+class CBlock_ThinRelay : public CBlock
```
I get why you used the underscore, but I think that we should stick with the standard bitcoin class naming convention, so in this case CBlockThinRelay.  The intention of these naming conventions is that capitalization and use of _ is hard to remember compared to the name itself.  With inconsistent naming, it becomes much harder to look things up from memory since most search tools aren't clever enough to find near matches.

gandrewstone commented on src/primitives/block.h:
```diff
@@ -75,8 +78,7 @@ class CBlock : public CBlockHeader
     // Xpress Validation: (memory only)
-    // However Orphans or Missing transactions that have been re-requested must be verifed
-    // because their inputs have never been checked.
+    // Orphans, or Missing transactions that have been re-requested, are stored here.
```
Let's start recovering doxygen auto-generated documentation.  This will save someone time later who has to go through all the code fixing everything. 

So please use //! to mark the comment so it is recognized as a docstring.

ptschip committed:
Naming change: CBlock_ThinRelay to CBlockThinRelay

ptschip commented:
ok, made the change  to CBlockThinRelay

On 19/03/2019 4:49 p.m., Andrew Stone wrote:
>
> *@gandrewstone* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/primitives/block.h
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1611#discussion_r267141107>:
>
> > @@ -193,6 +195,26 @@ class CBlock : public CBlockHeader
>      uint64_t GetBlockSize() const;
>  };
>  
> +// Used for thin type blocks that we want to reconstruct
> +class CBlock_ThinRelay : public CBlock
>
> I get why you used the underscore, but I think that we should stick
> with the standard bitcoin class naming convention, so in this case
> CBlockThinRelay. The intention of these naming conventions is that
> capitalization and use of _ is hard to remember compared to the name
> itself. With inconsistent naming, it becomes much harder to look
> things up from memory since most search tools aren't clever enough to
> find near matches.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1611#pullrequestreview-216473618>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0Ng4NqXA-Ng4Gh36hNsA6rQxub8tks5vYXdtgaJpZM4b6bF3>.
>

sickpig commented:
@ptschip need to be rebased.

ptschip committed:
Use Doxygen commenting for CBlockThinRelay

ptschip commented:
done
On 19/03/2019 4:52 p.m., Andrew Stone wrote:
>
> *@gandrewstone* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/primitives/block.h
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1611#discussion_r267141739>:
>
> > @@ -75,8 +78,7 @@ class CBlock : public CBlockHeader
>  
>  public:
>      // Xpress Validation: (memory only)
> -    // However Orphans or Missing transactions that have been re-requested must be verifed
> -    // because their inputs have never been checked.
> +    // Orphans, or Missing transactions that have been re-requested, are stored here.
>
> Let's start recovering doxygen auto-generated documentation. This will
> save someone time later who has to go through all the code fixing
> everything.
>
> So please use //! to mark the comment so it is recognized as a docstring.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1611#pullrequestreview-216474351>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0M_kHMDXRI5Jp5b4HU-brfyxAx0Bks5vYXgkgaJpZM4b6bF3>.
>

ptschip commented:
done

On 20/03/2019 2:22 a.m., Andrea Suisani wrote:
>
> @ptschip <https://github.com/ptschip> need to be rebased.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1611#issuecomment-474750209>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0EjjJVhebqnWT76Q7FqgMGerUk_aks5vYf3igaJpZM4b6bF3>.
>

ptschip commented:
rebased
